uint256.go

```diff
@@ -1358,33 +1355,23 @@ var (
 // Log10 returns the log in base 10, floored to nearest integer.
 // **OBS** This method returns '0' for '0', not `-Inf`.
 func (z *Int) Log10() uint {
-	// For some bit-lengths, there's only one possible value. Example:
-	// three bits can only represent [100 ... 111], or [4 ... 7]
-	// Ergo, bitlen:3 -> log10 == 0
-	res := lut[z.BitLen()%257]
-	if res >= 0 {
-		return uint(res)
-	}
-	// It was negative, which is a signal that we need to do one more check
-	// do determine which log it is. First remove the negation
-	res = -res
-
-	// We now lookup via the power of tens. Example:
-	// bitlen 4, [1000 ... 1111], or [8 .. 15]
-	// In order to figure out if it is '0' or '1', we only need to do one comparison,
-	// is it larger or smaller than '10'?
-
-	// For bitlengths < 20, we can use the uint64-space
-	if res < 20 {
-		// Uint64-space
-		if z.CmpUint64(pows64[res]) < 0 {
-			return uint(res - 1)
-		}
-		return uint(res)
+	// The following algorithm is taken from "Bit twiddling hacks"
+	// https://graphics.stanford.edu/~seander/bithacks.html#IntegerLog10
+	//
+	// The idea is that log10(z) = log2(z) / log2(10)
+	// log2(z) trivially is z.Bitlen()
+	// 1/log2(10) is a constant ~ 1233 / 4096. The approximation is correct up to 5 digit after
+	// the decimal point and it seems no further refinement is needed.
+	// Our tests check all boundary cases anyway.
+
+	bitlen := z.BitLen()
+	if bitlen == 0 {
+		return 0
```

Comment:
Or we may check `t != 0` case in the main `if`, what do you think?

Reply:
> The function comment says we return 0 for 0

Doesn't it? I thought it did

Reply:
> We don't really need this special case, since the code below it works equally well for the zero-case.

That statement was wrong. I don't know, might be better to put it back the way it was. That big clause is complicated enough as it is :)

Reply:
Ok, fixed.